### PR TITLE
[Feature] Implement file cache

### PR DESCRIPTION
**Pull request by zachfeldman**
Building on https://github.com/Shopify/erb-lint/pull/191, this attempts to finish up the implementation of a file cache. Also adds associated tests for `cli` class and new tests for `cache` class.

### CODE REVIEW

File: lib/erb_lint/cache.rb
Comment (ChrisBr) [inline lib/erb_lint/cache.rb:12]: Not sure how this is handled in the rest of the library but maybe we should pass in an io or logger so we could silence these outputs?

Something like

```ruby
def initialize(config, file_loader = nil, logger = STDOUT)
  # ...
  @logger = logger
  @logger.puts "Cache mode is on"
end
```

File: lib/erb_lint/cache.rb
Comment (ChrisBr) [inline lib/erb_lint/cache.rb:85]: I wonder if we should put the library version into the cache key too?

File: lib/erb_lint/cli.rb
Comment (ChrisBr) [inline lib/erb_lint/cli.rb:48]: This sentence doesn't sound like correct english? 

```suggestion
          failure!("cache directory doesn't exist, skipped deletion.")
```

File: lib/erb_lint/cli.rb
Comment (ChrisBr) [inline lib/erb_lint/cli.rb:149]: I wonder if we can also be a bit more smart here? Something what Rubocop does with max files in cache and then we prune the oldest when it's getting higher.

**Review by ChrisBr — COMMENTED** [review]
Looks great to me 👍

File: lib/erb_lint/cli.rb
Comment (rafaelfranca) [inline lib/erb_lint/cli.rb:43]:
```suggestion

      if clear_cache?
```

File: lib/erb_lint/cli.rb
Comment (rafaelfranca) [inline lib/erb_lint/cli.rb:119]:
```suggestion

      if with_cache? && !autocorrect?
```

File: lib/erb_lint/cli.rb
Comment (rafaelfranca) [inline lib/erb_lint/cli.rb:124]:
```suggestion

      log_offense_stats(runner, filename)
```

File: lib/erb_lint/cli.rb
Comment (rafaelfranca) [inline lib/erb_lint/cli.rb:129]: This method is never called with `autocorrect?` so we don't need to check again.
```suggestion
      if cache.include?(filename)
```

File: lib/erb_lint/cache.rb
Comment (rafaelfranca) [reply lib/erb_lint/cache.rb:85]: We should.

**Comment by zachfeldman** [comment]
Thanks for your comments @ChrisBr and @rafaelfranca ! I'm hoping to address them one night this evening or this weekend since I'm working on this PR in my spare time.

File: lib/erb_lint/cache.rb
Comment (zachfeldman) [reply lib/erb_lint/cache.rb]: Good call. Added a commit to do that.

File: lib/erb_lint/cli.rb
Comment (zachfeldman) [reply lib/erb_lint/cli.rb]: Yeah not sure how that ended up there, probably some copy pasta - fixed in the latest update.

File: lib/erb_lint/cli.rb
Comment (zachfeldman) [reply lib/erb_lint/cli.rb:152]: We definitely could but, naively, you can just run `erblint <files> --with-cache --prune-cache` every time and the code as written will remove outdated cache files only. Maybe we can leave more smart cache pruning for a future PR?

File: lib/erb_lint/cli.rb
Comment (zachfeldman) [reply lib/erb_lint/cli.rb]: Good catch! I removed it in the latest update.

**Comment by zachfeldman** [comment]
@ChrisBr @rafaelfranca thanks for the comments, I think I've replied to all of them. Are either of you able to approve me to run the CI workflow against my branch?

**Comment by zachfeldman** [comment]
@rafaelfranca I think the tests should be fixed, I believe they were failing on CI because the checksum was slightly different there vs locally. I mocked out some of the methods it uses to generate the checksum instead and everything is passing for me locally still.

**Review by ChrisBr — APPROVED** [review]
Nice 👏

**Comment by zachfeldman** [comment]
@rafaelfranca or @etiennebarrie are you able to run CI for me? I have some time today to fix things up if there are still failures 😃

**Comment by zachfeldman** [comment]
Rebased against master, specs were passing before the rebase but just want to stay up to date.

**Comment by ChrisBr** [comment]
@rafaelfranca do you want to have another look and 🚢 ?

**Review by eterry1388 — CHANGES_REQUESTED** [review]
While I haven't properly given this a review, in my testing I found that if there were a large number of errors, the caching actually greatly slows down the run rather than makes it faster. Also, it consumes a lot more memory. My assumption is that the cache files are getting loaded into memory and staying there until the end of the run. I bet it's taking a lot of time to parse through these massive cache files, a lot more time than just parsing through the erb files without caching in the first place. You can see my testing results here: https://github.com/Shopify/erb-lint/issues/158#issuecomment-1259495326

**Comment by zachfeldman** [comment]
Hey @eterry1388 , good find I appreciate the information! Do you have a repo of ERB files I could use to try to reproduce this? I could attempt to make the cache more performant in this scenario if so, or try to reproduce it myself, but writing a script to mass modify files to make errors doesn't really seem to be a representative test.

I do agree that it's a _little_ edge-casey, because I think in the real world most people who'd enable the cache would have been using `erb-lint` for long enough to have most files compliant and only a few to actually lint each time that had been changed. But if you were starting from scratch, using the cache, and had a lot of corrections to make, this behavior could be surprising. At the very least I could add a code comment or something to the README about this to warn people.

**Comment by eterry1388** [comment]
@zachfeldman I just took the repository I work on at my day job and added more checks to it. Turns out we were not compliant with most of the new cops I added. So I didn't do a mass change to make the failures. Maybe try doing something similar to that in your own "real-world" repository? Sorry I don't have a public repo I can share with you.

**Comment by Schwad** [comment]
Hey folks I haven't been workin on this but will have a look from my end monday too ;)

File: lib/erb_lint/cli.rb
Comment (etiennebarrie) [inline lib/erb_lint/cli.rb:137]: Could we avoid the `File.exist?` then `File.read` by removing `Cache#include?` and relying on `Cache#get` returning `nil` if there's no file? That skips one system call, we just open the file, instead of doing a `fstat` then `open`.

**Review by etiennebarrie — COMMENTED** [review]
I'm seeing an error with the DeprecatedClasses linter when trying to load it from cache:

```
Forwarding the following positional argument failed: #<ERBLint::Linters::DeprecatedClasses::RuleSet:0x00007f6035ec2dd0>
.bundle/gems/smart_properties-1.17.0/lib/smart_properties.rb:149:in `rescue in initialize'
.bundle/gems/smart_properties-1.17.0/lib/smart_properties.rb:146:in `initialize'
src/github.com/Shopify/erb-lint/lib/erb_lint/linter_config.rb:18:in `new'
src/github.com/Shopify/erb-lint/lib/erb_lint/linter_config.rb:18:in `block (2 levels) in to_array_of'
src/github.com/Shopify/erb-lint/lib/erb_lint/linter_config.rb:18:in `map'
src/github.com/Shopify/erb-lint/lib/erb_lint/linter_config.rb:18:in `block in to_array_of'
.bundle/gems/smart_properties-1.17.0/lib/smart_properties/property.rb:70:in `instance_exec'
.bundle/gems/smart_properties-1.17.0/lib/smart_properties/property.rb:70:in `convert'
.bundle/gems/smart_properties-1.17.0/lib/smart_properties/property.rb:92:in `prepare'
.bundle/gems/smart_properties-1.17.0/lib/smart_properties/property.rb:123:in `set'
.bundle/gems/smart_properties-1.17.0/lib/smart_properties.rb:139:in `block in initialize'
.bundle/gems/smart_properties-1.17.0/lib/smart_properties/property_collection.rb:55:in `block in each'
.bundle/gems/smart_properties-1.17.0/lib/smart_properties/property_collection.rb:55:in `each'
.bundle/gems/smart_properties-1.17.0/lib/smart_properties/property_collection.rb:55:in `each'
.bundle/gems/smart_properties-1.17.0/lib/smart_properties.rb:135:in `initialize'
src/github.com/Shopify/erb-lint/lib/erb_lint/linter_config.rb:33:in `initialize'
src/github.com/Shopify/erb-lint/lib/erb_lint/offense.rb:42:in `new'
src/github.com/Shopify/erb-lint/lib/erb_lint/offense.rb:42:in `from_json'
src/github.com/Shopify/erb-lint/lib/erb_lint/cache.rb:17:in `block in get'
```

File: lib/erb_lint/cli.rb
Comment (zachfeldman) [reply lib/erb_lint/cli.rb]: Yeah, I tried to avoid it, and this lead me to explore a few other optimizations and places that we can avoid repeatedly `read`-ing or `stat`-ing the file. For instance, when calculating the checksum, we again read the file if you look at the source of `file` on `Digest::SHA1` class:
https://ruby-doc.org/stdlib-2.4.0/libdoc/digest/rdoc/Digest/Instance.html#method-i-file

this also unfortunately led to the demise of the `#[]=` syntax for the getter but I think it's worth it for clarity, unless you can think of other ways to do this.

Check out the latest commit for the changes!

File: lib/erb_lint/cli.rb
Comment (zachfeldman) [reply lib/erb_lint/cli.rb]: Sure, I agree that it's sort of unrelated to the `CLI` class - I instead put this logic internally to the `Cache` class and run prune operations on a `cache.close` method call from `CLI`, so CLI isn't aware of pruning, only of the places it has to call the cache class at the beginning and end of linting in a recent commit I just pushed. How's that?

**Comment by zachfeldman** [comment]
> I'm seeing an error with the DeprecatedClasses linter when trying to load it from cache:
> 
> ```
> Forwarding the following positional argument failed: #<ERBLint::Linters::DeprecatedClasses::RuleSet:0x00007f6035ec2dd0>
> .bundle/gems/smart_properties-1.17.0/lib/smart_properties.rb:149:in `rescue in initialize'
> .bundle/gems/smart_properties-1.17.0/lib/smart_properties.rb:146:in `initialize'
> src/github.com/Shopify/erb-lint/lib/erb_lint/linter_config.rb:18:in `new'
> src/github.com/Shopify/erb-lint/lib/erb_lint/linter_config.rb:18:in `block (2 levels) in to_array_of'
> src/github.com/Shopify/erb-lint/lib/erb_lint/linter_config.rb:18:in `map'
> src/github.com/Shopify/erb-lint/lib/erb_lint/linter_config.rb:18:in `block in to_array_of'
> .bundle/gems/smart_properties-1.17.0/lib/smart_properties/property.rb:70:in `instance_exec'
> .bundle/gems/smart_properties-1.17.0/lib/smart_properties/property.rb:70:in `convert'
> .bundle/gems/smart_properties-1.17.0/lib/smart_properties/property.rb:92:in `prepare'
> .bundle/gems/smart_properties-1.17.0/lib/smart_properties/property.rb:123:in `set'
> .bundle/gems/smart_properties-1.17.0/lib/smart_properties.rb:139:in `block in initialize'
> .bundle/gems/smart_properties-1.17.0/lib/smart_properties/property_collection.rb:55:in `block in each'
> .bundle/gems/smart_properties-1.17.0/lib/smart_properties/property_collection.rb:55:in `each'
> .bundle/gems/smart_properties-1.17.0/lib/smart_properties/property_collection.rb:55:in `each'
> .bundle/gems/smart_properties-1.17.0/lib/smart_properties.rb:135:in `initialize'
> src/github.com/Shopify/erb-lint/lib/erb_lint/linter_config.rb:33:in `initialize'
> src/github.com/Shopify/erb-lint/lib/erb_lint/offense.rb:42:in `new'
> src/github.com/Shopify/erb-lint/lib/erb_lint/offense.rb:42:in `from_json'
> src/github.com/Shopify/erb-lint/lib/erb_lint/cache.rb:17:in `block in get'
> ```

I haven't been able to look into this failure yet. I'd like to create a spec that runs each linter type with the cache on to catch exceptions like this, or maybe I can just raise an exception if people try to use these linters with the cache on. Hopefully I'll have some more time this long weekend!

**Comment by zachfeldman** [comment]
@etiennebarrie @eterry1388 @Schwad I think this is ready for another review. I changed how the cache works, storing only the amount of information needed for the output of each `Offense` rather than the `Offense` itself. I _think_ this will fix the problem with the `DeprecatedClasses` linter but I wasn't able to get it set up correctly in my project - a representative test for `DeprecatedClasses` I can run would be appreciated! These changes should at least majorly increase performance as we won't have to serialize and deserialize the entirety of each `Offense` object and inject the source into these objects which was taking up massive amounts of RAM. Instead I introduced a simplified `CachedOffense` object with only the information needed to display each offense. 

I think I also addressed your other code comments @etiennebarrie  about opening too many files and where the pruning logic should live!

File: lib/erb_lint/cached_offense.rb
Comment (eterry1388) [inline lib/erb_lint/cached_offense.rb:35]: `severity` is sometimes `nil`.

```suggestion
        parsed_json[:severity]&.to_sym
```

**Review by eterry1388 — APPROVED** [review]

**Review by etiennebarrie — COMMENTED** [review]
I'm not seeing the exceptions I reported last time.

We do have cops that error, and when you use the cache you don't get the error anymore, but I'm not sure we need to do something about it.

Tiny feedback about the name: I wish it was `--cache` rather than `--with-cache`. Also should we enable pruning by default? What's the reason for not pruning?

**Comment by zachfeldman** [comment]
@etiennebarrie great! As mentioned before the cache is opt-in, so I'm not too concerned about that - subsequent pull requests can be made if people aren't happy with what information is stored/restored for their specific linter/use case, but I think this is far beyond a basic skeleton to add to for this feature at this point. I think I'm about ~25 hours of work in.

I changed the name to `cache` (this was a holdover from the first implementation https://github.com/Shopify/erb-lint/pull/191/files#diff-1156d85dde58090796995118a6420fe59feb9f48234b5f550f3b9ed2b2ea23a6R131). I also enabled pruning by default - I agree that it makes sense to do so, it just adds slightly more risk to those running the tool for the first time, but it'll probably cause even more problems later on to _not_ have it enabled by default.

I also added a README. Do you have any other concerns I can address? Thanks for your time.

**Comment by etiennebarrie** [comment]
> I also enabled pruning by default

For comparison, I checked what Rubocop does, it prunes not on the files that didn't have a hit but on the total number of cache files, and removes the oldest ones. But it must make it possible to use caching while linting a single file without pruning all the other cached results (locally in dev typically).

> it just adds slightly more risk to those running the tool for the first time

How so?

> Do you have any other concerns I can address?

* One thing I noticed is that we use the ability to tweak the Rubocop cache directory. It's not so much of a concern, more like a  new feature that can be added in another PR.

* I don't think we should display "Skipping deletion of new cache result" or "Cleaning deleted cached file with checksum". It just adds to the output without much value.

---

Other than that I tested it on our app and it works well.

**Comment by zachfeldman** [comment]
Great thanks @etiennebarrie ! I haven't had time yet to get back to look at those suggestions like I thought I would this weekend (man, the autumn is really beautiful here in New York so it's hard to want to be on my laptop) but I might some night this week or next weekend. 

I wasn't clear on if these last comments are enough for you to withhold your approval/merging of this to `main`, or if these are just nice to haves - I think we just need **you specifically** to approve this to merge it at this point, since we have 2 other approvals and you're a maintainer, right? Meanwhile maybe we can merge this so more people can test it and make some of your suggestions in follow up PRs since this one is already pretty huge (perhaps indicated by the fact that there's already been a PR merged to this PR :))? Unless there's anything else in it that makes you uncomfortable/worried about a merge.

**Comment by etiennebarrie** [comment]
> I wasn't clear on if these last comments are enough for you to withhold your approval

I'm just isn't clear about the risk that pruning adds?

> > it just adds slightly more risk to those running the tool for the first time
> 
> How so?

And I don't like the output of all files pruned


> * I don't think we should display "Skipping deletion of new cache result" or "Cleaning deleted cached file with checksum". It just adds to the output without much value.

The rest can be tackled later IMO.

**Comment by zachfeldman** [comment]
Great @etiennebarrie , see https://github.com/Shopify/erb-lint/pull/268/commits/e08285533ad14d9ffcfc17ae1475f7d416647452 to reduce the output of the pruning process. Ideally there'd be some "verbose" mode for erb-lint I could optionally hook that output into but again I'm just trying to get to a merge at this point.

Regarding the pruning - my point about risk was mostly a general one. All of this code is pretty new, anything we can do to reduce the risk of those using it is good IMHO. Sometimes it's better to just try to do less - ideally I'd want to put out the `--cache` and `--prune` options and let people experiment with them before running them both at the same time. But a few people have tried this branch and seemingly not reported any issues I'm not thinking of, and the code is well unit tested, so it's probably fine. As I keep pointing out all of this should be opt-in anyway and not affect anyone not using the cache.

**Comment by etiennebarrie** [comment]
I think we're good! Thanks a lot. I think we'll want another PR to make the directory configurable and then we can ship a new version.

**Pull request merged**
